relay-cardinality/src/redis/limiter.rs

```diff
@@ -55,39 +55,48 @@ impl RedisSetLimiter {
     /// Returns an iterator over all entries which have been accepted.
     fn check_limits(
         &self,
-        con: &mut Connection,
-        state: &LimitState<'_>,
-        scope: QuotaScoping,
-        entries: Vec<RedisEntry>,
+        connection: &mut Connection<'_>,
+        state: &mut LimitState<'_>,
         timestamp: UnixTimestamp,
-    ) -> Result<CheckedLimits> {
+    ) -> Result<Vec<CheckedLimits>> {
         let limit = state.limit;
 
-        metric!(
-            histogram(CardinalityLimiterHistograms::RedisCheckHashes) = entries.len() as u64,
-            id = state.id(),
-        );
+        let scopes = state.take_scopes();
 
-        let keys = scope
-            .slots(timestamp)
-            .map(|slot| scope.into_redis_key(slot));
-        let hashes = entries.iter().map(|entry| entry.hash);
+        let mut num_hashes: u64 = 0;
 
-        // The expiry is a off by `window.granularity_seconds`,
-        // but since this is only used for cleanup, this is not an issue.
-        let result = self
-            .script
-            .invoke(con, limit, scope.redis_key_ttl(), hashes, keys)?;
+        let mut pipeline = self.script.pipe();
+        for (scope, entries) in &scopes {
+            let keys = scope
+                .slots(timestamp)
+                .map(|slot| scope.into_redis_key(slot));
+
+            let hashes = entries.iter().map(|entry| entry.hash);
+            num_hashes += hashes.len() as u64;
+
+            // The expiry is a off by `window.granularity_seconds`,
+            // but since this is only used for cleanup, this is not an issue.
+            pipeline.add_invocation(limit, scope.redis_key_ttl(), hashes, keys);
+        }
 
         metric!(
-            histogram(CardinalityLimiterHistograms::RedisSetCardinality) = result.cardinality,
+            gauge(CardinalityLimiterGauges::RedisCheckHashes) = num_hashes,
             id = state.id(),
         );
 
-        Ok(CheckedLimits {
-            entries,
-            statuses: result.statuses,
-        })
+        let results = pipeline.invoke(connection)?;
+
+        scopes
+            .into_iter()
+            .zip(results)
```

Comment:
Does the pipeline always return a response for each invocation?

Reply:
Yes! I'll add some docs.